src/codegen/generate.ts

```diff
@@ -908,6 +964,56 @@ export default class ApiGenerator {
     return this.opts?.optimistic ? callOazapftsFunction("ok", [ex]) : ex;
   }
 
+  /**
+   * Does three things:
+   * 1. Add a `metadata` property.
+   * 2. Record parent schemas in `this.parentSchemaRefs`. A parent schema refers to a schema that
+   *    has a `discriminator` property which is neither used in conjunction with `oneOf` nor
+   *    `anyOf`.
+   * 3. Make all mappings of parent schemas explicit to generate types immediately.
+   */
+  preprocessSchemas(schemas: {
+    [key: string]: OpenAPIV3.ReferenceObject | SchemaObject;
+  }) {
+    const prefix = "#/components/schemas/";
+
+    for (const name of Object.keys(schemas)) {
+      const schema = schemas[name];
+      if (isReference(schema)) continue;
+      schema.metadata = { selfRef: prefix + name };
+      if (!schema.discriminator || !!schema.oneOf || !!schema.anyOf) continue;
+      this.parentSchemaRefs.add(prefix + name);
+    }
+
+    const isExplicit = (
+      discriminator: OpenAPIV3.DiscriminatorObject,
+      ref: string,
+    ) => {
+      const refs = Object.values(discriminator.mapping || {});
+      return refs.includes(ref);
+    };
+
+    for (const name of Object.keys(schemas)) {
+      if (this.parentSchemaRefs.has(prefix + name)) continue;
+      const schema = schemas[name];
+      if (isReference(schema)) continue;
+      if (!schema.allOf) continue;
+      for (const schema_ of schema.allOf) {
+        if (!isReference(schema_)) continue;
+        if (!this.parentSchemaRefs.has(schema_.$ref)) continue;
+        const parentSchema = schemas[
+          getRefBasename(schema_.$ref)
+        ] as SchemaObject;
+        const discriminator = parentSchema.discriminator!;
+        if (isExplicit(discriminator, prefix + name)) continue;
+        if (!discriminator.mapping) {
+          discriminator.mapping = {};
+        }
+        discriminator.mapping[name] = prefix + name;
+      }
+    }
+  }
```

Comment:
I find this part very hard to scan and to reason about.

Would it be possible to maybe break it up a bit and/or add more comments?